nflog/nflog.go

```diff
@@ -351,27 +408,37 @@ func (l *nlog) GC() (int, error) {
 
 // Query implements the Log interface.
 func (l *nlog) Query(params ...QueryParam) ([]*pb.Entry, error) {
-	q := &query{}
-	for _, p := range params {
-		if err := p(q); err != nil {
-			return nil, err
+	start := time.Now()
+	l.metrics.queriesTotal.Inc()
+
+	entries, err := func() ([]*pb.Entry, error) {
+		q := &query{}
+		for _, p := range params {
+			if err := p(q); err != nil {
+				return nil, err
+			}
+		}
+		// TODO(fabxc): For now our only query mode is the most recent entry for a
+		// receiver/group_key combination.
+		if q.recv == nil || q.groupKey == nil {
+			// TODO(fabxc): allow more complex queries in the future.
+			// How to enable pagination?
+			return nil, errors.New("no query parameters specified")
 		}
-	}
-	// TODO(fabxc): For now our only query mode is the most recent entry for a
-	// receiver/group_key combination.
-	if q.recv == nil || q.groupKey == nil {
-		// TODO(fabxc): allow more complex queries in the future.
-		// How to enable pagination?
-		return nil, errors.New("no query parameters specified")
-	}
 
-	l.mtx.RLock()
-	defer l.mtx.RUnlock()
+		l.mtx.RLock()
+		defer l.mtx.RUnlock()
```

Comment:
There are 3 returns points below where we'd have to repeat that. A defer has an overhead of something at the order of 20-60ns with a 2x improvement coming with go1.8.

Say we send out 10 notifications per second, which would even overload an organization with 1000 people on-call this will give us <0.0000005 of that one second being used for the defer. I'd argue this is fine for a load estimate about 100x above normal usage.